proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,364 @@
+length of the history array exceeds the maximum depth (512), the oldest entries
+will be removed from the array (i.e., "popped").
+
+The `root_hash` is the hash of all the `EpochRewardReserve` accounts hashes.
+Initially, at the start of the epoch's first block, the `root_hash` is set to
+`None`. Once all the `EpochRewardReserve` accounts are filled and the reward
+computation phase is complete, the `root_hash` is calculated by hashing all the
+`EpochRewardReserve` hashes, which is defined below.
+
+```
+root_hash = Hash(EpochRewardReserve[..].reserve_hash)
+
+
+
+                     +-------------+
+                     |             |
+                     |  root_hash  |
+                     |             |
+                     +------+------+
+                            |
+    +--------------+--------+-----------+
+    |              |                    |
++---+----+   +-----+--+             +---+-----+
+|reserve |   | reserve|   ......    | reserve |
+| hash   |   |  hash  |             |  hash   |
++--------+   +--------+             +---------+
+
+```
+
+### `EpochRewardReserve` Sysvar Account
+
+`EpochRewardReserve` sysvar accounts track the rewards to be distributed for each
+individual block in the `reward distribution phase`.
+
+The address of `EpochRewardReserve` account for a particular block is specified
+by PDA as follows
+
+```
+fn get_epoch_reward_reserve_addr(epoch: u64, distribution_phase_index: u64)
+   -> (Pubkey, u8)
+{
+   let (address, bump_seed) = Pubkey::find_program_address(&[
+        b"inflation-rewards-distributor",
+        b"stake",
+        &epoch.to_le_bytes(),
+        &distribution_phase_index.to_le_bytes(),
+    ],
+    &sysvar::id(),
+   );
+   (address, bump_seed)
+}
+```
+
+where `distribution_phase_index` is the block height offset since the start of
+reward distribution phase.
+
+
+```
+struct EpochRewardReserve {
+   reserve_hash: Hash,
+   reward_balance: u64,
+}
+```
+
+The account balance of `EpochRewardsReserve` is equal to the total amount (in the
+`AccountInfo::lamports` field) of all rewards to be distributed in a block. And
+the `reward_balance` field shadow the `AccountInfo::lamports`. The
+`reserve_hash` is computed from the rewards in the block as follows.
+
+```
+reserve_hash = hash([(stake_account_key, reward_amount)])
+```
+
+At the end of reward computation phase, `M` `EpochRewardsReserve` sysvar
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks. Each of the `M` blocks is responsible for distributing
+the reward from one partition of the rewards stored in the `EpochRewardReserve`
+account with the address specified by the `get_epoch_reward_reserve_addr`
+function above. Once all rewards have been distributed, the balance of the
```

Comment:
Is there a reason for all of the different accounts? If it's about storing intermediate hashes, you could more easily store the entire list of hashes and rewards in one account.

Reply:
I think the main idea behind this was belts and suspenders safety. When the new lamports are minted at the boundary, if they were all held in a single account there's the greater risk of a bug that accidentally distributes too many lamports to some stake accounts. In the multiple account case, perhaps only 4095 stake accounts are affected. 

But I'm not sure that's a super strong argument tbh, and I'm failing to come up with another reason do to it this way (probably @HaoranYi has one though?)

Reply:
I thought through the different possible error scenarios, and how 1 account vs many accounts handle the situations.

* The total reward calculation is correct, but the individual distributions are incorrect. With one account that stores all of the intermediate hashes, you'll fail to calculate the correct hash when you do the incorrect distributions, so you'll fail correctly. With N accounts, you'll also fail.
* The total rewards calculation is correct, the individual distributions are incorrect, but somehow the hash is correct. In this case, one account would certainly get all of the rewards wrong, but most likely this problem would also impact all N accounts and still mess up all rewards.
* The total rewards calculation is incorrect, and the individual distributions are incorrect. You're screwed no matter what here.

I think the intermediate hash provides the extra safety that's needed and simplifies the model a lot. Looking through these error cases, I'm having a hard time finding a situation in which N accounts provides more safety, since a lot of these error situations are catastrophic no matter what. Do you have a specific situation in mind where N accounts helps?

Reply:
N accounts isn't about safety, it's about progress tracking.  if we emit one account for each block in the calculation phase and close one account for each block in the distribution phase, you can always recover the progress of the entire rewards period based on the population of `EpochRewardReserve` accounts alone

Reply:
Maybe I'm missing something, but using the account population seems much more complicated than storing a list of expected hashes and current hash all in one account.

Reply:
> Maybe I'm missing something, but using the account population seems much more complicated than storing a list of expected hashes and current hash all in one account.

but that doesn't get us a snapshot-modification-free solution. if we're starting from a snapshot created during the rewards period, we can restart the rewards period from scratch and use the `EpochRewardsReserve` account population to checkpoint our work until we've caught up to the snapshot block's expected progress, at which point we join the rest of the cluster in block-by-block checkpointing

Reply:
Sorry, I still think I'm missing something here. If you start from a snapshot created during rewards distribution, which is what I'm assuming you meant, you can't restart from scratch because some of the `EpochRewardsReserve` accounts will no longer exist. One of the goals of the proposal is to be able to pick off from where you left off, so why are you mentioning restarting distribution from scratch?

I'm imagining just one account with:

```
struct EpochRewardReserve {
    current_hash: Hash,
    all_hashes: Vec<EpochReward>,
}

struct EpochReward {
   reserve_hash: Hash,
   reward_balance: u64,
}
```

So this way, if you restart during a snapshot, you know exactly where to pick up based on this account -- check for where `current_hash == all_hashes[i].reserve_hash`

Reply:
I am working on an updated proposal. 
No more async reward computation in the background.
Rewards is going to be compuated in the first block. 
Therefore, handling snapshot and restarting will not be needed. 

If I understand it correctly, the current open questions is whether we should just store the full rewards directly instead of using the chucked `EpochRewardReserve` SysVar accounts. 
@joncinque I can see your augment for the benefit of storing the full rewards directly. I will add this as an open question in the proposal.